Add registration through invite

andrezafeu opened the pull request:
[Accept User Invitations](https://trello.com/c/bomO7FOx/13-accept-user-invitations)


## Prerequisite PRs

Generate Invite Tokens to Users (https://github.com/PocketLobby/constituents/pull/35) (merged)

## Dependent PRs
None

## Notes for Reviewer
Expected:
- If invite token is present, H1 header should update with the first name of the person that invited you
If registration is successful:
- Invite token param from `/register` path, if present, should be saved in db as `invited_by`. Otherwise, `invited_by` should be nil.
- New user should see his/her invite url in success page.

e91096 commented on frontend/components/RegistrationForm.jsx:
```diff
@@ -11,6 +11,9 @@ import {
 } from '../actions/userActions';
 
+const match = window.location.href.match(/register\/([a-zA-Z\d-]+)/);
+const invitedBy = match ? match[1] : null;
```
i think it's a better pattern to save these kinds of things to component state

e91096 commented on frontend/components/RegistrationForm.jsx:
```diff
@@ -108,6 +115,10 @@ class RegistrationForm extends Component {
           errors.push('State should be two letters.');
         }
         break;
+      case 'invitedBy':
```
we don't need this case bc we catch `invitedBy` above in line 99!

e91096 replied:
ah okay i understand. i'm not sure if it makes sense here as this is more for user-input validation errors. question: does a user need to have an invite token to register?

andrezafeu replied:
I agree that we should allow them to register. I rebased and updated the code to validate the invited_by only in the back-end

e91096 commented on frontend/components/RegistrationForm.jsx:
```diff
@@ -121,7 +132,11 @@ class RegistrationForm extends Component {
+            <h1>Join Pocket Lobby</h1>
+          ) : ( 
+            <h1>{invitedBy.charAt(0).toUpperCase() + invitedBy.substr(1).match('^[^-]+')} has invited you to join Pocket Lobby</h1>
+          )}
```
nice!

e91096 commented on app/models/user.rb:
```diff
@@ -33,4 +36,10 @@ def generate_invite_token
 
+  def validate_invited_by
+    return if self.invited_by == nil
+    # REVIEW regex to make sure entire string matches
```
is this still todo?

andrezafeu replied:
oops, that's done

e91096 commented on frontend/components/RegistrationForm.jsx:
```diff
@@ -11,6 +11,7 @@ import {
   registerUser,
 } from '../actions/userActions';
 
+
```
extra space

andrezafeu replied:
removed : )

e91096 commented on frontend/components/RegistrationForm.jsx:
```diff
@@ -86,10 +99,11 @@ class RegistrationForm extends Component {
       state: 'State',
       USCitizen: 'US Citizen',
       zip: 'Zip',
+      invitedBy: 'Invited By',
```
not needed

e91096 commented on frontend/routes.jsx:
```diff
@@ -22,7 +22,7 @@ const routes = [
   { path: /^\/login/, action: () => <LoginForm /> },
   { path: /^\/register$/, action: () => <RegistrationForm /> },
-  { path: /^\/register\/([a-zA-Z\d-]+)$/, action: () => <RegistrationForm /> },
+  { path: /^\/register\/(.+)$/, action: () => <RegistrationForm /> },
```
why not make it `(.*)` and just have one path that routes to the `RegistrationForm`?

andrezafeu replied:
I tried but this didn't work for me. I'm getting an error in the async function in the router, but can't figure out what's wrong with it. Do you mind taking a look at that when you review this?

e91096 commented on frontend/components/SuccessStub.jsx:
```diff
@@ -7,6 +7,9 @@ import { logoutUser } from '../actions/userActions';
+        Thanks for registering!
+      </h1>
       <h2>
         Your invite URL is {window.location.origin}/register/{props.currentUser.invite_token}
```
i'm not seeing an invite-token after i register (as a new user)

andrezafeu replied:
fixed

e91096 replied:
works!

andrezafeu commented on frontend/components/RegistrationForm.jsx:
```diff
@@ -30,6 +30,17 @@ class RegistrationForm extends Component {
 
+  componentDidMount() {
+    const href = window.location.href;
+    if (href.includes('inviteToken')) {
```
It will be true if the inviteToken is present in the url. Otherwise, it will be false. But I just noticed that I don't need this `if` statement. That's what you had in mind when you made this question? I refactored the code : )

e91096 replied:
hmmm i mean is it ever the case that the url contains the string `'inviteToken'`? the one that we present to the user looks like this: `http://localhost:3000/register/test-user-iswjp3mian` and not `http://localhost:3000/register/inviteToken=test-user-iswjp3mian`

andrezafeu commented:
@e91096 regarding the invite token: before I was also getting the invite token from the url, but I was using `.match`. For some reason, that piece of code stopped working when I rebased the branch. So I updated it to this one.
For the success page I think that, since the invite token used is generated in the BE, it makes more sense to serve from the BE than from the url. The url would look 'cleaner' and it doesn't let the user manipulate the value. Let me know if you agree!

e91096 commented:
yup that's a great idea!
i was more concerned that the url we provide to the users doesn't match the format of the url we're expecting to parse when we're in the registration form?

andrezafeu commented:
@e91096 Ohhhhhh I see now. I misunderstood your comment before!

andrezafeu commented:
@e91096 when I defined the route in the FE, I thought it would be better to accept any content after the `register/` part of the path. So if the url gets broken in any way, this won't prevent the user from being registered. It's more a less what you suggested [here](https://github.com/PocketLobby/constituents/pull/44#discussion_r217927889) . If there's any change you think it's worth doing, could you be more specific? Or were you just double checking that the route works? Let me know!
